java-squid/src/main/java/org/sonar/java/cfg/CFG.java

```diff
@@ -0,0 +1,602 @@
+        // process condition
+        currentBlock = createBranch(s, bodyBlock, falseBranch);
+        buildCondition(s.condition(), bodyBlock, falseBranch);
+        loopback.successors.add(currentBlock);
+        currentBlock = createBlock(currentBlock);
+        break;
+      }
+      case DO_STATEMENT: {
+        DoWhileStatementTree s = (DoWhileStatementTree) tree;
+        Block falseBranch = currentBlock;
+        Block loopback = createBlock();
+        // process condition
+        currentBlock = createBranch(s, loopback, falseBranch);
+        buildCondition(s.condition(), loopback, falseBranch);
+        // process body
+        currentBlock = createBlock(currentBlock);
+        continueTargets.addLast(loopback);
+        breakTargets.addLast(falseBranch);
+        build(s.statement());
+        breakTargets.removeLast();
+        continueTargets.removeLast();
+        loopback.successors.add(currentBlock);
+        currentBlock = createBlock(currentBlock);
+        break;
+      }
+      case FOR_EACH_STATEMENT: {
+        ForEachStatement s = (ForEachStatement) tree;
+        //TODO(npe) One solution is to create a forstatement node depending on type of expression (iterable or array) and build CFG from it.
+        Block afterLoop = currentBlock;
+        Block loopback = createBlock();
+        continueTargets.addLast(loopback);
+        breakTargets.addLast(afterLoop);
+        build(s.statement());
+        breakTargets.removeLast();
+        continueTargets.removeLast();
+        build(s.variable());
+        build(s.expression());
+        break;
+      }
+      case FOR_STATEMENT: {
+        ForStatementTree s = (ForStatementTree) tree;
+        Block falseBranch = currentBlock;
+        // process step
+        currentBlock = createBlock();
+        Block stepBlock = currentBlock;
+        for (StatementTree updateTree : Lists.reverse(s.update())) {
+          build(updateTree);
+        }
+        // process body
+        currentBlock = createBlock(currentBlock);
+        continueTargets.addLast(stepBlock);
+        breakTargets.addLast(falseBranch);
+        build(s.statement());
+        breakTargets.removeLast();
+        continueTargets.removeLast();
+        Block body = currentBlock;
+        // process condition
+        ExpressionTree condition = s.condition();
+        if (condition != null) {
+          currentBlock = createBranch(s, body, falseBranch);
+          buildCondition(condition, body, falseBranch);
+        } else {
+          currentBlock = createUnconditionalJump(s, body);
+        }
+        stepBlock.successors.add(currentBlock);
+        // process init
+        currentBlock = createBlock(currentBlock);
+        for (StatementTree init : Lists.reverse(s.initializer())) {
+          build(init);
+        }
+        break;
+      }
+      case TRY_STATEMENT: {
+        //FIXME only path with no failure constructed for now, (not taking try with resources into consideration).
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Take the required action to fix the issue indicated by this comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1134)